### ISSUE TITLE
cross-compiling broken in 0.37.0

### ISSUE THREAD

**Issue by msink** [mesonbuild/meson#1223]
The Meson build system
Version: 0.37.0
Source dir: F:\avian\hello-c
Build dir: F:\avian\hello-c\build\linux-arm-gcc-64-release
Build type: cross build
Project name: hello
Native c compiler: gcc (gcc 5.4.0)
Cross c compiler: c:/cygwin-cross/bin/aarch64-unknown-linux-gnu-gcc (gcc 4.9.0)
Host machine cpu family: arm
Host machine cpu: aarch64
Target machine cpu family: arm
Target machine cpu: aarch64
Build machine cpu family: x86_64
Build machine cpu: x86_64
Traceback (most recent call last):
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\mesonmain.py", line 289, in run
    app.generate()
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\mesonmain.py", line 177, in generate
    intr.run()
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreter.py", line 2215, in run
    super().run()
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 124, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 145, in evaluate_codeblock
    raise e
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 139, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 160, in evaluate_statement
    return self.evaluate_if(cur)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 216, in evaluate_if
    self.evaluate_codeblock(node.elseblock)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 145, in evaluate_codeblock
    raise e
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 139, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 152, in evaluate_statement
    return self.assignment(cur)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 547, in assignment
    value = self.evaluate_statement(node.value)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 150, in evaluate_statement
    return self.function_call(cur)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreterbase.py", line 371, in function_call
    return self.funcs[func_name](node, self.flatten(posargs), kwargs)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\interpreter.py", line 1726, in func_dependency
    dep = dependencies.find_external_dependency(name, self.environment, kwargs)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\dependencies.py", line 1386, in find_external_dependency
    raise pkg_exc
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\dependencies.py", line 1375, in find_external_dependency
    pkgdep = PkgConfigDependency(name, environment, kwargs)
  File "c:\program files (x86)\python35-32\lib\site-packages\mesonbuild\dependencies.py", line 116, in __init__
    if 'pkgconfig' not in env.cross_info.config['binaries']:
NameError: name 'env' is not defined

**Comment by msink** [comment]
after #1242

The Meson build system
Version: 0.38.0.dev1
Source dir: F:\avian\hello-c
Build dir: F:\avian\hello-c\build\linux-arm-gcc-32-debug
Build type: cross build
Project name: hello
Native c compiler: gcc (gcc 5.4.0)
Cross c compiler: c:/cygwin-cross/bin/armv7hl-unknown-linux-gnueabi-gcc (gcc 4.8.2)
Host machine cpu family: arm
Host machine cpu: armv7hl
Target machine cpu family: arm
Target machine cpu: armv7hl
Build machine cpu family: x86_64
Build machine cpu: x86_64

Meson encountered an error in file meson.build, line 10, column 2:
Pkg-config not found.

**Comment by QuLogic** [comment]
That's fixed in #1221.

**Comment by msink** [comment]
after #1224 + #1221:

```
The Meson build system
Version: 0.38.0.dev1
Source dir: F:\avian\hello-c
Build dir: F:\avian\hello-c\build\osx-clang-64-release
Build type: cross build
Project name: hello
Native c compiler: gcc (gcc 5.4.0)
Cross c compiler: c:\cygwin-cross\bin\x86_64-apple-darwin14-clang (clang 3.8.1)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Build machine cpu family: x86_64
Build machine cpu: x86_64
Cross dependency zlib found: YES 1.2.8

Executing subproject bin2c.

Project name: bin2c
Native c compiler: gcc (gcc 5.4.0)
Cross c compiler: c:\cygwin-cross\bin\x86_64-apple-darwin14-clang (clang 3.8.1)
Build targets in project: 1

Subproject bin2c finished.
Build targets in project: 2
Traceback (most recent call last):
  File "F:\avian\meson\mesonbuild\mesonmain.py", line 289, in run
    app.generate()
  File "F:\avian\meson\mesonbuild\mesonmain.py", line 179, in generate
    g.generate(intr)
  File "F:\avian\meson\mesonbuild\backend\ninjabackend.py", line 185, in generate
    self.generate_rules(outfile)
  File "F:\avian\meson\mesonbuild\backend\ninjabackend.py", line 745, in generate_rules
    self.generate_compile_rules(outfile)
  File "F:\avian\meson\mesonbuild\backend\ninjabackend.py", line 1530, in generate_compile_rules
    self.generate_llvm_ir_compile_rule(compiler, True, outfile)
  File "F:\avian\meson\mesonbuild\backend\ninjabackend.py", line 1386, in generate_llvm_ir_compile_rule
    ' '.join(self.get_cross_info_lang_args(compiler, is_cross)),
  File "F:\avian\meson\mesonbuild\backend\ninjabackend.py", line 1406, in get_cross_info_lang_args
    return self.environment.cross_info.config['properties'][lang + '_args']
TypeError: unsupported operand type(s) for +: 'ClangCCompiler' and 'str'
```

```
[host_machine]
system     = 'darwin'
cpu_family = 'x86_64'
cpu        = 'x86_64'
endian     = 'little'

[binaries]
c          = 'c:\\cygwin-cross\\bin\\x86_64-apple-darwin14-clang'
cpp        = 'c:\\cygwin-cross\\bin\\x86_64-apple-darwin14-clang++'
ar         = 'c:\\cygwin-cross\\bin\\x86_64-apple-darwin14-ar'
strip      = 'c:\\cygwin-cross\\bin\\x86_64-apple-darwin14-strip'
pkgconfig  = 'c:\\cygwin-cross\\bin\\x86_64-apple-darwin14-pkg-config'

[properties]
root       = 'c:\\cygwin-cross\\usr\\SDK\\MacOSX10.10.sdk'
```

**Comment by ignatenkobrain** [comment]
```diff
diff --git a/mesonbuild/backend/ninjabackend.py b/mesonbuild/backend/ninjabackend.py
index 6a6782e6..a20a35fd 100644
--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -1383,7 +1383,7 @@ rule FORTRAN_DEP_HACK
             return
         rule = 'rule llvm_ir{}_COMPILER\n'.format('_CROSS' if is_cross else '')
         args = [' '.join([ninja_quote(i) for i in compiler.get_exelist()]),
-                ' '.join(self.get_cross_info_lang_args(compiler, is_cross)),
+                ' '.join(self.get_cross_info_lang_args(compiler.language, is_cross)),
                 ' '.join(compiler.get_output_args('$out')),
                 ' '.join(compiler.get_compile_only_args())]
         if mesonlib.is_windows():
```

@msink can you try this patch?

**Comment by msink** [comment]
Yes, after last patch cross-build works as in 0.36

Release build for osx sthill fails:
```
c:\cygwin-cross\bin\x86_64-apple-darwin14-clang @hello.rsp
ld: unknown option: -O1
clang-3.8: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

But this is another issue :)

**Comment by msink** [comment]
In master branch (4317edc) my cross-build works - this issue resolved.

**Comment by nirbheek** [comment]
Closing as fixed.
